### DISCUSSION TITLE
Listing all installed models?

### DISCUSSION THREAD

**Discussion by BramVanroy** [explosion/spaCy#7625] [Help: Other Questions] [answered]
I am writing code that involves initialising a parser. Users must be able to simply use language codes. Since `spacy.load` does not work with language codes any more, I am wondering whether there is a way to programmatically list the models that have been installed?

If I remember correctly, in v2 we were required to use `pip list` to find models - but doing that programmatically was error-prone because the relationship to spaCy was not clear and you were never sure whether `en_potatoes_small` was a spaCy models or a different library. I cannot find this discussion anymore, however. Has anything changed in v3 in that respect? Is it now easier to list models, and perhaps even find their attached language code?

Ideally, I'd have a function where I can get a list of installed models per language, or at least a list of installed models (assuming that the prefix is always the language code, so I can create a mapping myself). Does something like this exist?

**Comment by polm** [comment, marked as answer]
There's actually a function for that in v3.

```
from spacy.util import get_installed_models
print(get_installed_models()) # list of model names as  strings
```

This is one of the nice effects of the [registry](https://spacy.io/api/top-level#registry).

**Reply by BramVanroy** [reply]
That's neat. I'll dig into how this works behind the scenes later. Especially curious how this interplays with the fact that models are installable packages.

For now, I am using something like this. There are probably cases where this does not work very well or selects the model that you may not want, but for now it does do the trick in my case.

```python
try:
    nlp = spacy.load(model_or_lang)
except OSError as exc:
    try:
        model_name = next(m for m in get_installed_models() if m.startswith(f"{model_or_lang}_"))
    except StopIteration:
        raise OSError(f"Could not find a spaCy model that is called '{model_or_lang}'"
                              f" or that starts with '{model_or_lang}_'") from exc

    logger.info(f"Specified spaCy model {model_or_lang} not found. Using {model_name} instead.")
    nlp = spacy.load(model_name)
```
